asciidoc/courses/llm-fundamentals/modules/3-intro-to-langchain/lessons/7-c-neo4j-vector-retriever/code/retriever_agent.py

```diff
@@ -27,61 +23,67 @@
     input_variables=["chat_history", "input"]
     )
 
-chat_chain = LLMChain(llm=llm, prompt=prompt, memory=memory, verbose=True)
+memory = ConversationBufferMemory(memory_key="chat_history", return_messages=True)
 
-embedding_provider = OpenAIEmbeddings(openai_api_key=OPENAI_API_KEY)
+chat_chain = LLMChain(llm=llm, prompt=prompt, memory=memory)
+
+youtube = YouTubeSearchTool()
 
 movie_plot_vector = Neo4jVector.from_existing_index(
     embedding_provider,
     url="bolt://localhost:7687",
     username="neo4j",
     password="pleaseletmein",
     index_name="moviePlots",
-    embedding_node_property="embedding", 
+    embedding_node_property="embedding",
     text_node_property="plot",
 )
 
-retrievalQA = RetrievalQA.from_llm(
-    llm=llm, 
+plot_retriever = RetrievalQA.from_llm(
+    llm=llm,
     retriever=movie_plot_vector.as_retriever(), 
     verbose=True, 
     return_source_documents=True
 )
 
 def run_retriever(query):
-    results = retrievalQA({"query":query})
-    return str(results)
+    results = plot_retriever.invoke({"query":query})
+    # format the results
+    movies = '\n'.join([doc.metadata["title"] + " - " + doc.page_content for doc in results["source_documents"]])
+    return movies
 
 tools = [
     Tool.from_function(
-        name="ChatOpenAI",
-        description="For when you need to chat about movies, genres or plots. The question will be a string. Return a string.",
+        name="Movie Chat",
```

Comment:
Do these need to be UpperCamelCase?

Reply:
Dont think so. They work fine with textual descriptions.

One of the changes I found in v 0.1 was unless I had "better" names for the tools the llm was less likely to pick the right tool based on the query.

Reply:
Thought I should check whether it was a convention to use UpperCamelCase. I dont think it is, the langchain documents have examples in "normal" text, CamelCase, kebab-case.